File: Bloom/Bloom/Services/VoiceProfileService.swift
import Foundation
import Supabase

final class VoiceProfileService {
    static let shared = VoiceProfileService()

    private init() {}

    struct VoiceProfileUpsert: Encodable {
        let user_id: UUID
        let sample_audio_url: String
    }

    struct VoiceProfile: Decodable {
        let id: UUID
        let user_id: UUID
        let sample_audio_url: String
        let created_at: String
        let updated_at: String
    }

    func uploadVoiceSample(userId: UUID, localFileURL: URL) async throws -> String {
        let fileName = "\(userId.uuidString)-voice-sample.m4a"
        let storagePath = "samples/\(fileName)"
        let data = try Data(contentsOf: localFileURL)

        try await SupabaseManager.shared.client.storage
            .from("voice-samples")
            .upload(
                storagePath,
                data: data,
                options: FileOptions(
                    contentType: "audio/m4a",
                    upsert: true
                )
            )

        let publicURL = try SupabaseManager.shared.client.storage
            .from("voice-samples")
            .getPublicURL(path: storagePath)

        let payload = VoiceProfileUpsert(
            user_id: userId,
            sample_audio_url: publicURL.absoluteString
        )

        try await SupabaseManager.shared.client
            .from("user_voice_profiles")
            .upsert(payload)
            .execute()

        return publicURL.absoluteString
    }

    func fetchVoiceProfile(userId: UUID) async throws -> VoiceProfile? {
        let response = try await SupabaseManager.shared.client
            .from("user_voice_profiles")
            .select()
            .eq("user_id", value: userId.uuidString)
            .limit(1)
            .execute()

        let rows = try JSONDecoder().decode([VoiceProfile].self, from: response.data)
        return rows.first
    }
}

File: Bloom/bloom-backend/app/services/voice_profile_service.py
from uuid import UUID
from app.db import supabase


def get_voice_profile(user_id: UUID):
    response = (
        supabase.table("user_voice_profiles")
        .select("*")
        .eq("user_id", str(user_id))
        .limit(1)
        .execute()
    )

    rows = response.data or []
    return rows[0] if rows else None


def update_provider_voice_id(user_id: UUID, provider_voice_id: str):
    response = (
        supabase.table("user_voice_profiles")
        .update({
            "provider": "elevenlabs",
            "provider_voice_id": provider_voice_id,
        })
        .eq("user_id", str(user_id))
        .execute()
    )

    rows = response.data or []
    return rows[0] if rows else None
